src/users/users.service.js

```diff
@@ -0,0 +1,55 @@
+import { Injectable, Dependencies } from '@nestjs/common';
+import { getModelToken } from '@nestjs/mongoose';
+
+@Injectable()
+@Dependencies(getModelToken('User'))
+export class UsersService {
+  constructor(userModel) {
+    this.userModel = userModel;
+  }
+
+  async list(offset, limit) {
+    return this.userModel
+      .find({})
+      .sort({ id: 1 })
+      .skip(offset)
+      .limit(limit)
+      .exec();
+  }
+
+  async create(id, { name, dob, address, description }) {
+    const user = this.userModel({
+      id,
+      name,
+      dob,
+      address,
+      description,
+    });
+    return user.save();
+  }
+
+  async getById(id) {
+    const user = await this.userModel.findOne({ id }).exec();
+    return user;
+  }
+
+  async update(id, { name, dob, address, description }) {
```

Comment:
if race condition, does this update method have issues?

Reply:
Yes, there may be some unexpected behaviors.

If we only take the `PUT` api or the `update` method into consideration:
Case 1. If the update requests come one by one, then the later request will overwrite the earlier request. 
Case 2. If two update requests come concurrently, then the changed fields will be merged, since mongoose only sends changed fields to the transaction command.

I think the behavior is not consistent, and may cause potential bugs.
One simple solution is to add a optimistic concurrency control (with a plugin like [mongoose-update-if-current](https://github.com/eoin-obrien/mongoose-update-if-current)), so that in Case 2 one of the request will be rejected. 
And the behavior will be simple and consistent.

If we take the whole `GET` and then `PUT` process into consideration:
1. Update request that finishes later will overwrite the request that finished earlier.

If the responsibility of user can be kept small, a `PUT` api is usually enough.
If we want to merge requests that each only changes part of the user, a `PATCH` api may be better.
If the responsibility of user grows to be bigger, we may need `PUT` api like `/users/:id/:subresource`.